### PR TITLE
search for compiler names in order

### LINKED ISSUE #518: Auto-detection of compilers with multiple names
Recently, I discovered that certain PGI compilers (`pgf77` and `pgf90`) were unable to build HDF4, while newer ones (`pgfortran`) had no trouble. I submitted PR #493 to solve this issue. Now, if your compilers.yaml contains something like:

``` yaml
pgi@15.10-0:
      cc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgcc
      cxx: /soft/pgi/15.10/linux86-64/15.10/bin/pgc++
      f77: /soft/pgi/15.10/linux86-64/15.10/bin/pgfortran
      fc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgfortran
```

HDF4 will build for you. However, when I run:

``` bash
spack compiler add /soft/pgi/15.10/linux86-64/15.10/bin
```

Spack still finds the older compilers first:

``` yaml
pgi@15.10-0:
      cc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgcc
      cxx: /soft/pgi/15.10/linux86-64/15.10/bin/pgCC
      f77: /soft/pgi/15.10/linux86-64/15.10/bin/pgf77
      fc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgf90
```

The [PGI Installation Guide](https://www.pgroup.com/doc/pgiinstall.pdf) recommends using the newer compilers, although the older ones are kept around for backwards compatibility. Although I could simply remove `pgCC`, `pgf77`, and `pgf90` from those lists, it would cause problems for anyone using older installations of PGI, which don't come with the newer compilers.

I propose we change the way that `spack compiler add` finds compilers. My suggestion would be to choose the first match it finds in `cxx_names`, `f77_names`, and `f90_names`. I believe the relevant code is in the `find` method of `lib/spack/spack/compiler.py`. Things start to get complicated since we also need to be able to find compilers with prefixes, suffixes, or version numbers, and much of the code is parallelized for efficiency.

If anyone would like to tackle this, and maybe wrap in a solution to any of the unresolved problems we've been seeing with Apple compilers, be my guest. Otherwise, I will hopefully get to it eventually. It doesn't significantly impact me since I can just `spack config edit compilers` and switch it over to the newer ones, but it might trip up a new Spack user.

### PR DESCRIPTION
**Pull request by becker33**
for compilers that have multiple executable names, search the names
in the priority order in `compilers/*.py` under \<lang\>_names to ensure
that newer names are caught before backwards compatibility names.

Fixes #518
Supersedes #6327

@adamjstewart

### CODE REVIEW

File: lib/spack/spack/compilers/__init__.py [hunk 1]
```diff
@@ -576,9 +576,8 @@ def _default(search_paths):
                         )
                         command_arguments.append(detect_version_args)
 
-        # Reverse it here so that the dict creation (last insert wins)
-        # does not spoil the intended precedence.
-        return reversed(command_arguments)
+        # Order already respecs priority from compiler class
```
Comment (adamjstewart) [inline lib/spack/spack/compilers/__init__.py]:
```suggestion
        # Order already respects priority from compiler class
```

File: lib/spack/spack/compilers/__init__.py [hunk 2]
```diff
@@ -576,9 +576,8 @@ def _default(search_paths):
                         )
                         command_arguments.append(detect_version_args)
 
-        # Reverse it here so that the dict creation (last insert wins)
-        # does not spoil the intended precedence.
-        return reversed(command_arguments)
+        # Order already respects priority from compiler class
```
Comment (scheibelp) [inline lib/spack/spack/compilers/__init__.py:579]: Generally when you remove a comment and replace it with a new one, the new comment shouldn't "respond" to the old one. In other words, I would rephrase like:

> Return command arguments using the order specified by the compiler class (i.e. leave them unchanged)

File: lib/spack/spack/test/cmd/compiler.py [hunk 3]
```diff
@@ -59,6 +69,81 @@ def mock_compiler_dir(tmpdir, mock_compiler_version):
     return str(tmpdir)
 
 
+@pytest.fixture()
+def mock_two_compiler_dirs(tmpdir, mock_compiler_version):
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/compiler.py:73]: Can the name be updated to include more detail about the issue:

`def mock_two_compilers_with_same_spec`

File: lib/spack/spack/test/cmd/compiler.py [hunk 4]
```diff
@@ -59,6 +69,81 @@ def mock_compiler_dir(tmpdir, mock_compiler_version):
...
+
+    gcc_path1.copy(gcc_path2, mode=True)
+    gcc_path1.copy(gxx_path2, mode=True)
+    gcc_path1.copy(gfortran_path2, mode=True)
+
+    return [str(tmpdir1), str(tmpdir2)]
+
+
+@pytest.fixture()
+def mock_pgi_dir(tmpdir, mock_pgi_version_string):
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/compiler.py:115]: Can the name include more details:

`def mock_pgi_old_and_new_fortran`

File: lib/spack/spack/test/cmd/compiler.py [hunk 5]
```diff
@@ -103,3 +188,47 @@ def test_compiler_add(
     new_compiler = new_compilers - old_compilers
     assert any(c.version == spack.version.Version(mock_compiler_version)
                for c in new_compiler)
+
+
+def test_compiler_add_multiple_in_path(
+    mock_two_compiler_dirs, mutable_empty_config, mock_compiler_version
+):
+    # Find compilers
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/compiler.py:196]: This `#` comment is unnecessary given that `compiler_find` is called right after. The test could use a docstring though (something like "When two compilers are detected with the same spec, make sure Spack uses the first one that appears in the ordered list of paths provided").

File: lib/spack/spack/test/cmd/compiler.py [hunk 6]
```diff
@@ -103,3 +188,47 @@ def test_compiler_add(
...
+    # Ensure new compiler is from the first path passed and has correct version
+    new_compilers = set(spack.compilers.all_compiler_specs())
+    new_comp = spack.compilers.compilers_for_spec('gcc@%s'
+                                                  % mock_compiler_version)
+    assert mock_two_compiler_dirs[0] in new_comp[0].cc
+    assert any(c.version == spack.version.Version(mock_compiler_version)
+               for c in new_compilers)
+
+
+def test_compiler_add_multiple_names(
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/compiler.py:214]: `def test_pgi_use_latest_fortran_name`?

File: lib/spack/spack/test/cmd/compiler.py [hunk 7]
```diff
@@ -103,3 +188,47 @@ def test_compiler_add(
...
+    mock_pgi_dir, mutable_empty_config, mock_pgi_version,
+    mock_pgi_version_string
+):
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/compiler.py:217]: Docstring, e.g.:

> Ensure that the `pgi` compiler returns "pgfortran" as the fortran compiler, preferring it over "pgf90" when both names are present (likewise for "pgf77")

**Review by scheibelp — CHANGES_REQUESTED** [review]
There is a sync issue

This looks great but I have several suggestions for naming/docstrings

**Comment by adamjstewart** [comment]
Is this ever going to be merged or should we just close this and #518? It's my oldest Spack issue, so I would love to see it disappear.

**Comment by becker33** [comment]
#17967 made the same change
